src/cmd-cloud-prune

```diff
@@ -0,0 +1,291 @@
+CACHE_MAX_AGE_METADATA = 60 * 5
+
+
+def main():
+    # Parse arguments and initialize variables
+    args = parse_args()
+    with open(BUILDFILES['sourceurl'], "r") as file:
+        builds_source_data_url = file.read()
+    bucket, prefix = parse_s3_url(builds_source_data_url)
+    cloud_config = get_cloud_config(args)
+    stream = args.stream
+
+    # Boto3 loads credentials from ~/.aws/config by default and we can change
+    # this default location by setting the AWS_CONFIG_FILE environment variable.
+    # The Python bindings don't support passing a config file.
+    # The alternative is to manually pass ACCESS_KEY and SECRET_KEY which isn't favourable.
+    if args.aws_config_file:
+        os.environ["AWS_CONFIG_FILE"] = args.aws_config_file
+    s3_client = boto3.client("s3")
+
+    # These lists are up to date as of schema hash
+    # 4c19aed3b3d84af278780bff63728510bb3e70613e4c4eef8cabd7939eb31bd8. If changing
+    # this hash, ensure that the list of supported and unsupported artifacts below
+    # is up to date.
+    supported = ["amis", "gcp"]
+    unsupported = ["aliyun", "azurestack", "digitalocean", "exoscale", "ibmcloud", "powervs", "azure"]
+
+    with open(args.policy, "r") as f:
+        policy = yaml.safe_load(f)
+    validate_policy(stream, policy)
+
+    with open(BUILDFILES['list'], "r") as f:
+        builds_json_data = json.load(f)
+
+    # Prune builds based on the policy
+    for action in ['cloud-uploads', 'images', 'build']:
+        if action not in policy[stream]:
+            continue
+        duration = get_period_in_months(policy[stream][action])
+        ref_date = datetime.datetime.now() - relativedelta(months=int(duration))
+
+        print(f"Pruning resources of type {action} older than {duration} months ({ref_date.date()}) on stream {stream}")
+        # Enumerating in reverse to go from the oldest build to the newest one
+        for index, build in enumerate(reversed(builds_json_data["builds"])):
+            build_id = build["id"]
+            if action in build.get("policy-cleanup", []):
+                print(f"Build {build_id} has already had {action} pruning completed")
+                continue
+            build_date = parse_fcos_version(build_id)
+
+            if build_date >= ref_date:
+                break
+            for arch in build["arches"]:
+                meta_prefix = os.path.join(prefix, f"builds/{build_id}/{arch}/meta.json")
+                meta_json = get_json_from_s3(s3_client, bucket, meta_prefix)
+                # Make sure the meta.json doesn't contain any cloud_platform that is not supported for pruning yet.
+                images = get_supported_images(meta_json, unsupported, supported)
+                current_build = Build(id=build_id, images=images, arch=arch)
+
+                match action:
+                    case "cloud-uploads":
+                        prune_cloud_uploads(current_build, cloud_config, build, builds_json_data, index, args.dry_run)
```

Comment:
I think there's a bit of a layer violation here. We're passing way too much state to this function. Passing both a `Build` and also `build`, *and* also `index`, *and* `builds_json_data` seems off.

I think let's have the `Build` tuple include the meta JSON. Then `prune_cloud_uploads` should just take `current_build`, `cloud_config`, and `args.dry_run`. Then here after the arch for-loop, update the `policy-cleanup` of `build` (this fixes a bug in the current code: we don't want to update the `policy-cleanup` key on the first iteration, only once we've done all arches).